lib/zz.mli

```diff
@@ -0,0 +1,183 @@
+     can safely discard [`Await] case (with [assert false]). *)
+
+  type dst = [ `Channel of out_channel | `Buffer of Buffer.t | `Manual ]
+  (** The type for output destinations. With a [`Manual] destination the client
+     must provide output storage with {!dst}. With [`String] or [`Channel]
+     destination the client can safely discard [`Flush] case (with [assert false]). *)
+
+  type encoder
+  (** The type for ZLIB encoders. *)
+
+  type ret = [ `Await of encoder | `End of encoder | `Flush of encoder ]
+
+  val encoder : src -> dst -> q:Dd.B.t -> w:window -> level:int -> encoder
+  (** [encoder src dst ~q ~w ~level] is an encoder that inputs from [src] and
+     that outputs to [dst].
+
+      {b Internal queue.}
+
+      [encoder] deals internally with compression algorithm and DEFLATE encoder.
+     To pass compression values to DEFLATE encoder, we need a queue [q]. Length
+     of it can be a bottleneck where a small one will let {!encode} to emit too
+     many [`Flush]. We recommend a queue as large as output buffer.
+
+      {b Window.}
+
+      ZLIB is able to constrain length of window used to do LZ77 compression.
+     However, small window can slow-down LZ77 compression algorithm. Small
+     window is mostly used to be able to inflate output into hostile
+     environment.
+
+      {b Level.}
+
+      Current implementation of ZLIB does not handle any compression level.
+     However, the client must give a level between 0 and 9. *)
+
+  val src_rem : encoder -> int
+  (** [src_rem e] is how many bytes it remains in given input buffer. *)
+
+  val dst_rem : encoder -> int
+  (** [dst_rem e] is how many bytes it remains in given output buffer. *)
+
+  val src : encoder -> bigstring -> int -> int -> encoder
+  (** [src e s j l] provides [e] with [l] bytes to read, starting at [j] in [s].
+     This byte range is read by calls to {!encode} with [e] until [`Await] is
+     returned. To signal the end of input call the function with [l = 0].
+
+     @raise Invalid_argument when [j] and [l] do not correspond to a valid
+     range. *)
+
+  val dst : encoder -> bigstring -> int -> int -> encoder
+  (** [dst e s j l] provides [e] with [l] bytes available to write, starting at
+     [j] in [s]. This byte range is read by calls to {!encode} with [e] until
+     [`Flush] is returned.
+
+     @raise Invalid_argument when [j] and [l] do not correspond to a valid
+     range. *)
+
+  val encode : encoder -> ret
+  (** [encode e0] is:
+
+      {ul
+      {- [`Await e1] if [e0] has a [`Manual] input source and awaits for more
+     input. The client must use {!src} with [e1] to provide it.}
+      {- [`Flush e1] if [e1] has a [`Manual] destination and needs more output
+     storage. The client must use {!dst} with [e1] to provide new buffer and
+     then call {!encode}.}
+      {- [`End e1] if [e1] encoded all input. Output buffer is possibly not
+     empty (it can be check with {!dst_rem}).}} *)
```

Comment:
"Output buffer is possibly not empty (it can be check with {!dst_rem})." is only about `Manual` source. `Buffer` and `Channel` cases takes care to flush output directly to the given buffer (with `Buffer.add_string`) or the `out_channel` (with `output_string`).

In others words, `dst_rem` should returns `0` if `dst` is `Buffer` or `Channel`.